### PR TITLE
[ci] Check contrib modules are using the correct core collector version

### LINKED ISSUE #8546: [CI] Check consistent version usage of core collector dependencies
A common issue (e.g. most recently found on the v0.47.0 release cycle) is for individual components that are not yet added to the main build to require a Collector version lower than the one on the main `go.mod` file. This makes it more difficult to bump the Collector version since, when we do that, it uncovers hidden issues.

We can build a small Go tool using the [modfile](https://pkg.go.dev/golang.org/x/mod/modfile) parser to walk over every `go.mod` file on the repository and check that the required version for `go.opentelemetry.io/collector/*` dependencies is the same as the one required in the main `go.mod` file.

### PR DESCRIPTION
**Pull request by hickeyma**
Description from @mx-psi in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8546#issuecomment-1084592613:

> My intention is to avoid the following situation:
> 1. Someone creates a PR adding a new examplemodule module, targeting Collector core vN module
> 2. Weeks pass, it gets reviewed and merged just before the vN+3 contrib release, still targeting vN
> 3. The release manager for vN+3 has then to manually update the examplemodule module code to be compatible with vN+3
>
> I want to avoid (3) and, instead, make the contributor adding examplemodule do that job before merging the PR, by comparing the Collector core version at the top-level go.mod file with the Collector core version at the examplemodule go.mod file.

**The result of check-collector-module-version (no incorrect versions):**

```console
There are 0 incorrect "go.opentelemetry.io/collector " version(s) in the module files.
There are 0 incorrect "go.opentelemetry.io/collector/model" version(s) in the module files.
```

**The result of check-collector-module-version (incorrect versions):**

```console
Incorrect version "v0.47.0" of "go.opentelemetry.io/collector " is included in "./receiver/couchbasereceiver/go.mod". It should be version "v0.48.0".
Incorrect version "v0.44.0" of "go.opentelemetry.io/collector " is included in "./exporter/awsemfexporter/go.mod". It should be version "v0.48.0".
There are 2 incorrect "go.opentelemetry.io/collector " version(s) in the module files.
```

**The result of check-collector-module-version (incorrect model versions):**

```console
There are 0 incorrect "go.opentelemetry.io/collector " version(s) in the module files.
Incorrect version "v0.43.0" of "go.opentelemetry.io/collector/model" is included in "./exporter/datadogexporter/go.mod". It should be version "v0.48.0".
There are 1 incorrect "go.opentelemetry.io/collector/model" version(s) in the module files.
```

Fixes #8546 

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

### CODE REVIEW

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (mx-psi) [inline .github/workflows/scripts/check-collector-module-version.sh:1]: Can you add a shebang?
```suggestion
#! /bin/bash

get_collector_version() {
```

Also, should we `set -euxo pipefail`?

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (mx-psi) [inline .github/workflows/scripts/check-collector-module-version.sh:47]: Shellcheck suggests doing:

```suggestion
COLLECTOR_MOD_VERSION=$(get_collector_version "$COLLECTOR_MODULE" "$MAIN_MOD_FILE")
```

for good argument hygiene

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (mx-psi) [inline .github/workflows/scripts/check-collector-module-version.sh:45]: I guess this works because you explicitly add a space after `COLLECTOR_MODULE`, is that right? If so, I would rather do `grep "$collector_module " "$mod_file"` and add a comment on the grep line documenting what the space is for.

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (hickeyma) [reply .github/workflows/scripts/check-collector-module-version.sh:1]: Sorry, forgot to round off the file before pushing. Done.

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (hickeyma) [reply .github/workflows/scripts/check-collector-module-version.sh:47]: Agree, done.

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (hickeyma) [reply .github/workflows/scripts/check-collector-module-version.sh:45]: That is probably a better place to put it. However, `check_collector_version_correct` is called using the `COLLECTOR_MODULE` and `COLLECTOR_MODEL_MODULE` strings. `COLLECTOR_MODEL_MODULE` does NOT have a space at end. For this reason, I think it will need to remain here in the initialisation of the string. I have added a comment. Let me know if that is ok.

**Comment by hickeyma** [comment]
Thanks for the review @mx-psi. dated and ready for review again.

**Review by mx-psi — APPROVED** [review]
Tested locally, seems to work fine. I would like another approver/maintainer to take a look before I mark as ready to merge

**Comment by mx-psi** [comment]
And thanks for doing this @hickeyma 🙇 I think this will make releases a bit smoother

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (jpkrohling) [inline .github/workflows/scripts/check-collector-module-version.sh:37]:
```suggestion
# Compare the collector main core version against all the collector component
```

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (jpkrohling) [inline .github/workflows/scripts/check-collector-module-version.sh:39]: This is checking all go.mod files against the main go.mod's, right? Perhaps using "versions" (plural) would be more appropriate for the function name?

```suggestion
check_collector_versions_correct() {
```

**Review by jpkrohling — COMMENTED** [review]
LGTM, just a couple of minor comments. Would you be able to add example outputs as a comment to this PR?

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (hickeyma) [reply .github/workflows/scripts/check-collector-module-version.sh:37]: Done

File: .github/workflows/scripts/check-collector-module-version.sh
Comment (hickeyma) [reply .github/workflows/scripts/check-collector-module-version.sh:39]: That makes sense. Updated.

**Comment by hickeyma** [comment]
Thanks for the review and feedback @jpkrohling. I have updated with your comments. It is ready for review again.

> Would you be able to add example outputs as a comment to this PR?

I added output to the description.

**Pull request merged**
